tests/test_user_profile.py

```diff
@@ -0,0 +1,151 @@
+                raise Exception("Oh no")
+
+        self.controller.storage = BadStorage()
+        problem = self.controller.get()
+        eq_(500, problem.status_code)
+        eq_(u"Oh no", problem.debug_message)
+        
+    def test_get_non_dictionary_profile_document(self):
+        """Test what happens if the profile_document is not a dictionary.
+        """
+        
+        class BadStorage(MockProfileStorage):
+            @property
+            def profile_document(self):
+                return u"Here it is!"
+
+        self.controller.storage = BadStorage()
+        problem = self.controller.get()
+        eq_(500, problem.status_code)
+        eq_(u"Profile profile_document is not a JSON object: u'Here it is!'.",
+            problem.debug_message)
+        
+    def test_get_non_dictionary_profile_document(self):
+        """Test what happens if the profile_document cannot be converted to JSON.
+        """
+        
+        class BadStorage(MockProfileStorage):
+            @property
+            def profile_document(self):
+                return dict(key=object())
+
+        self.controller.storage = BadStorage()
+        problem = self.controller.get()
+        eq_(500, problem.status_code)
+        assert problem.debug_message.startswith(
+            u"Could not convert profile document to JSON: {'key': <object object"
+        )
+
+    def test_put_bad_media_type(self):
+        """You must send the proper media type with your PUT request."""
+        headers = {"Content-Type" : "application/json"}
+        body = json.dumps(dict(settings={}))
+        problem = self.controller.put(headers, body)
+        eq_(415, problem.status_code)
+        eq_('Expected vnd.librarysimplified/user-profile+json',
+            problem.detail)
+
+    def test_put_invalid_json(self):
+        """You can't send any random string that's not JSON."""
+        headers = {"Content-Type" : ProfileController.MEDIA_TYPE}
+        problem = self.controller.put(headers, "blah blah")
+        eq_(400, problem.status_code)
+        eq_(u"Submitted profile document was not valid JSON.", problem.detail)
+
+    def test_put_non_object(self):
+        """You can't send any random JSON string, it has to be an object."""
+        headers = {"Content-Type" : ProfileController.MEDIA_TYPE}
+        problem = self.controller.put(headers, json.dumps("blah blah"))
+        eq_(400, problem.status_code)
+        eq_(u'Submitted profile document was not a JSON object.',
+            problem.detail)
+        
+    def test_attempt_to_set_read_only_setting(self):
+        """You can't change the value of a setting that's not
+        writable.
+        """
+        headers = {"Content-Type" : ProfileController.MEDIA_TYPE}
+        body = json.dumps(dict(settings=dict(key="new value")))
+        problem = self.controller.put(headers, body)
+        eq_(400, problem.status_code)
+        eq_('"key" is not a writable setting.', problem.detail)
```

Comment:
I'm not sure what this comment means. I'm testing that the name of the key ('key', defined on line 136) is used in the error message. If 'key' is a confusing name for a key I can change it.